### PR TITLE
Refactor extras and movie multi-part parsing

### LINKED ISSUE #6906: [Issue]: Slow performances when scanning large amount of files inside a directory
### Please describe your bug

On my Debian 10.11 Server Jellyfin 10.7.7 installed from APT repo I get a slow web interface when there is any library scanning in progress.
There is the same problem on two servers I5 8GB RAM and on a Intel Core i9-9900K up to 5GHz and 64GB DDR4 and RTX2080
Library on multiples Seagate Exos 16TB drives and about 4000 movies and 150 TVshow
When I add 1 movie the scanning take very long, I don't understand why Jellyfin run entire library scan for one movie added or renamed.
There is no way to make GUI more responsive during scan ? This is not a CPU power problem but I don't understand why...

### Jellyfin Version

10.7.7

### if other:

_No response_

### Environment

```markdown
- OS: Debian 10.11
- Virtualization: No
- Clients: 1 during test
- Browser: Any (last firefox stable, last chromium stable, last chrome stable...)
- FFmpeg Version: The one installed with Jellyfin stable Debian package
- Playback Method:
- Hardware Acceleration: RTX2080, problem occur also without any hardware acceleration enabled
- Plugins: problem occur also without plugins installed, but I use Reports and TMDb Box Set
- Reverse Proxy: Apache/2.4.38 (Debian) with SSL enabled
- Base URL: /jellyfin
- Networking: Gigabit Ethernet and Gigabit FTTH
- Storage: 100TB of Seagate Exos 16TB drives
```

### Jellyfin logs

```shell
No error showed, it's a performance issue. please ask if more information required
```

### FFmpeg logs

```shell
GUI thread performance issue when library scanning in progress
```

### Please attach any browser or client logs here

No JS error

### Please attach any screenshots here

_No response_

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by cvium**
**Changes**
Tried to consolidate the extras parsing instead of having multiple methods doing the same things.

Removed support for extras in flat file structures (they will be outright ignored).

**Issues**
Fixes #6906

### CODE REVIEW

File: Emby.Naming/Video/StackResolver.cs [hunk 1]
```diff
@@ -82,15 +73,20 @@ public IEnumerable<FileStack> ResolveAudioBooks(IEnumerable<AudioBookFileInfo> f
...
             var list = files
-                .Where(i => i.IsDirectory || VideoResolver.IsVideoFile(i.FullName, _options) || VideoResolver.IsStubFile(i.FullName, _options))
+                .Where(i => i.IsDirectory || VideoResolver.IsVideoFile(i.FullName, namingOptions) || VideoResolver.IsStubFile(i.FullName, namingOptions))
                 .OrderBy(i => i.FullName)
+                .Select(f => (f.IsDirectory, FileName: GetFileNameWithExtension(f), f.FullName))
                 .ToList();
 
-            var expressions = _options.VideoFileStackingRegexes;
+            // TODO is there a "nicer" way?
+            var cache = new Dictionary<(string, Regex, int), Match>();
```
Comment (cvium) [inline Emby.Naming/Video/StackResolver.cs:87]: It's not super great, but the regex allocates a lot of strings (O(N^N))

File: MediaBrowser.LocalMetadata/Images/LocalImageProvider.cs [hunk 2]
```diff
@@ -106,28 +106,29 @@ private static IEnumerable<FileSystemMetadata> GetFiles(BaseItem item, bool incl
...
 
             // Exit if the cache dir does not exist, alternative solution is to create it, but that's a lot of empty dirs...
             if (!Directory.Exists(path))
             {
-                return Enumerable.Empty<FileSystemMetadata>();
+                yield break;
             }
 
-            if (includeDirectories)
+            var files = directoryService.GetFileSystemEntries(path).OrderBy(i => Array.IndexOf(BaseItem.SupportedImageExtensions, i.Extension ?? string.Empty));
```
Comment (cvium) [inline MediaBrowser.LocalMetadata/Images/LocalImageProvider.cs:120]: These changes are somewhat unrelated to the extras, but related to the slowness of #6906

File: Emby.Server.Implementations/Library/Resolvers/BaseVideoResolver.cs [hunk 3]
```diff
@@ -49,120 +49,71 @@ public override T Resolve(ItemResolveArgs args)
...
+            if (videoType.HasValue)
+            {
+                video.VideoType = videoType.Value;
             }
+            else
+            {
+                SetVideoType(video, videoInfo);
+            }
+
+            Set3DFormat(video, videoInfo);
```
Comment (1337joe) [inline Emby.Server.Implementations/Library/Resolvers/BaseVideoResolver.cs:114]: Does moving this out of the loop also fix #6609?

File: tests/Jellyfin.Naming.Tests/Video/VideoListResolverTests.cs [hunk 4]
```diff
@@ -265,7 +283,10 @@ public void TestStackedWithTrailer2()
                 }).ToList(),
                 _namingOptions).ToList();
 
-            Assert.Single(result);
+            Assert.Equal(3, result.Count);
```
[the remark this answers is not in the document]
Comment (cvium) [reply tests/Jellyfin.Naming.Tests/Video/VideoListResolverTests.cs:286]: None

File: Emby.Server.Implementations/Library/Resolvers/BaseVideoResolver.cs (hunk 3 above)
Comment (cvium) [reply Emby.Server.Implementations/Library/Resolvers/BaseVideoResolver.cs:114]: Doubt it

**Review by crobibero — APPROVED** [review]

File: Emby.Naming/Video/ExtraResolver.cs [hunk 5]
```diff
@@ -76,7 +60,7 @@ public ExtraResult GetExtraInfo(string path)
                 {
                     var filename = Path.GetFileName(path);
 
-                    var regex = new Regex(rule.Token, RegexOptions.IgnoreCase);
+                    var regex = new Regex(rule.Token, RegexOptions.IgnoreCase | RegexOptions.Compiled);
```
Comment (IDisposable) [inline Emby.Naming/Video/ExtraResolver.cs:63]: Would be better to have the `Regex` cached in the `rule` so we don't keep `new`ing them up at `IsMatch` time.

Comment (cvium) [reply Emby.Naming/Video/ExtraResolver.cs:63]: That is true. The rules are due for a general refactor, which is out of scope of this (already big) PR.

File: Emby.Naming/Video/StackResolver.cs (hunk 1 above)
Comment (cvium) [reply Emby.Naming/Video/StackResolver.cs:87]: BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.1348 (2004/?/20H1)
AMD Ryzen 5 3600, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=6.0.100
  [Host]     : .NET Core 6.0.0 (CoreCLR 6.0.21.52210, CoreFX 6.0.21.52210), X64 RyuJIT
  DefaultJob : .NET Core 6.0.0 (CoreCLR 6.0.21.52210, CoreFX 6.0.21.52210), X64 RyuJIT


|  Method |     Mean |   Error |  StdDev |     Gen 0 |    Gen 1 |    Gen 2 | Allocated |
|-------- |---------:|--------:|--------:|----------:|---------:|---------:|----------:|
|   Cache | 146.7 ms | 2.69 ms | 3.40 ms | 1000.0000 | 750.0000 | 250.0000 |   7.71 MB |
| NoCache | 142.0 ms | 1.91 ms | 1.70 ms | 1000.0000 |        - |        - |   9.83 MB |

File: Emby.Naming/Video/FileStackRule.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    private readonly Regex _tokenRegex;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="FileStackRule"/> class.
+    /// </summary>
+    /// <param name="token">Token.</param>
+    /// <param name="isNumerical">Whether the file stack rule uses numerical or alphabetical numbering.</param>
+    public FileStackRule(string token, bool isNumerical)
+    {
+        _tokenRegex = new Regex(token, RegexOptions.IgnoreCase);
```
Comment (IDisposable) [inline Emby.Naming/Video/FileStackRule.cs:20]: Worth doing `.Compiled` here?

File: Emby.Naming/Video/FileStackRule.cs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    public bool Match(string input, [NotNullWhen(true)] out (string StackName, string PartType, string PartNumber)? result)
+    {
+        result = null;
+        var match = _tokenRegex.Match(input);
+        if (!match.Success)
+        {
+            return false;
+        }
+
+        var partType = match.Groups["parttype"].Success ? match.Groups["parttype"].Value : "vol";
```
Comment (IDisposable) [inline Emby.Naming/Video/FileStackRule.cs:44]: Curious what `"vol"` represents...

File: Emby.Naming/Video/FileStackRule.cs (hunk 6 above)
Comment (cvium) [reply Emby.Naming/Video/FileStackRule.cs:20]: NamingOptions is a Singleton so it doesn't matter much

File: Emby.Naming/Video/FileStackRule.cs (hunk 7 above)
Comment (cvium) [reply Emby.Naming/Video/FileStackRule.cs:44]: Probably should've been "unknown" but what matters is that it's different from the other regexes

File: Emby.Naming/Video/ExtraResolver.cs (hunk 5 above)
Comment (cvium) [reply Emby.Naming/Video/ExtraResolver.cs:63]: Just to follow up, there's currently no Regex-based ExtraTypeRule implemented

File: Emby.Naming/Video/StackResolver.cs [hunk 8]
```diff
@@ -12,45 +12,36 @@ namespace Emby.Naming.Video
...
 
         /// <summary>
         /// Resolves audiobooks from paths.
         /// </summary>
         /// <param name="files">List of paths.</param>
         /// <returns>Enumerable <see cref="FileStack"/> of directories.</returns>
-        public IEnumerable<FileStack> ResolveAudioBooks(IEnumerable<AudioBookFileInfo> files)
+        public static IEnumerable<FileStack> ResolveAudioBooks(IEnumerable<AudioBookFileInfo> files)
         {
             var groupedDirectoryFiles = files.GroupBy(file => Path.GetDirectoryName(file.Path));
```
[the remark this answers is not in the document]
Comment (cvium) [reply Emby.Naming/Video/StackResolver.cs:45]: There's no place to save it and the enumeration only occurs once. LINQ operations like GroupBy pull the entire collection before doing the actual grouping, so there should be no extra penalty.

**Review by crobibero — APPROVED** [review]

File: Emby.Naming/Video/ExtraResolver.cs [hunk 9]
```diff
@@ -102,5 +86,66 @@ public ExtraResult GetExtraInfo(string path)
...
+            for (var pos = files.Count - 1; pos >= 0; pos--)
+            {
+                var current = files[pos];
+                // ignore non-extras and multi-file (can this happen?)
+                if (current.ExtraType == null || current.Files.Count > 1)
+                {
+                    continue;
+                }
+
+                var currentFile = files[pos].Files[0];
```
Comment (Bond-009) [inline Emby.Naming/Video/ExtraResolver.cs:114]:
```suggestion
                var currentFile = current.Files[0];
```

**Review by Bond-009 — CHANGES_REQUESTED** [review]
I *think* your usage of `RegexOptions.Compiled` is wrong
https://docs.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-options#compiled-regular-expressions

**Review by Bond-009 — APPROVED** [review]

**Review by 1337joe — APPROVED** [review]

**Pull request merged**